File: src/main/go/leetcode/editor/cn/GroupAnagrams.go
//49.group-anagrams
//给你一个字符串数组，请你将 字母异位词 组合在一起。可以按任意顺序返回结果列表。
//
// 字母异位词 是由重新排列源单词的所有字母得到的一个新单词。
//
//
//
// 示例 1:
//
//
//输入: strs = ["eat", "tea", "tan", "ate", "nat", "bat"]
//输出: [["bat"],["nat","tan"],["ate","eat","tea"]]
//
// 示例 2:
//
//
//输入: strs = [""]
//输出: [[""]]
//
//
// 示例 3:
//
//
//输入: strs = ["a"]
//输出: [["a"]]
//
//
//
// 提示：
//
//
// 1 <= strs.length <= 10⁴
// 0 <= strs[i].length <= 100
// strs[i] 仅包含小写字母
//
//
// Related Topics 数组 哈希表 字符串 排序 👍 1525 👎 0

package main

import (
	"fmt"
	"sort"
	"strconv"
	"strings"
)

// leetcode submit region begin(Prohibit modification and deletion)
func groupAnagrams(strs []string) [][]string {
	mp := map[string][]string{}
	for _, str := range strs {
		key := std2(str)
		mp[key] = append(mp[key], str)
	}
	//make(Type, len, cap)
	ans := make([][]string, 0, len(mp))
	for _, v := range mp {
		ans = append(ans, v)
	}
	return ans

}

func std(str string) string {
	s := []byte(str)
	sort.Slice(s, func(i, j int) bool { return s[i] < s[j] })
	return string(s)
}

func std2(str string) string {
	cnt := [26]int{}
	for _, b := range str {
		cnt[b-'a']++
	}
	build := strings.Builder{}
	for i, i2 := range cnt {
		if i2 > 0 {
			build.WriteString(string(i + 'a'))
			if i2 > 1 {
				build.WriteString(strconv.Itoa(i2))
			}
		}
	}
	return build.String()
}

//leetcode submit region end(Prohibit modification and deletion)

// test from here
func main() {

	//fmt.Println(groupAnagrams([]string{"eat", "tea", "tan", "ate", "nat", "bat"}))
	fmt.Println(std("eatttttttt"))
	fmt.Println(std2("eatttttttt"))
}

File: src/main/python/leetcode/editor/en/GroupAnagrams.py
# 49.group-anagrams
"""
Given an array of strings strs, group the anagrams together. You can return the 
answer in any order. 

 
 Example 1: 

 
 Input: strs = ["eat","tea","tan","ate","nat","bat"] 
 

 Output: [["bat"],["nat","tan"],["ate","eat","tea"]] 

 Explanation: 

 
 There is no string in strs that can be rearranged to form "bat". 
 The strings "nat" and "tan" are anagrams as they can be rearranged to form 
each other. 
 The strings "ate", "eat", and "tea" are anagrams as they can be rearranged to 
form each other. 
 

 Example 2: 

 
 Input: strs = [""] 
 

 Output: [[""]] 

 Example 3: 

 
 Input: strs = ["a"] 
 

 Output: [["a"]] 

 
 Constraints: 

 
 1 <= strs.length <= 10⁴ 
 0 <= strs[i].length <= 100 
 strs[i] consists of lowercase English letters. 
 

 Related Topics Array Hash Table String Sorting 👍 21635 👎 742

"""
from collections import defaultdict
from typing import *


# leetcode submit region begin(Prohibit modification and deletion)
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        def key(s: str):
            arr = [0] * 26
            for c in s:
                # cannot use like java: c-'a'
                arr[ord(c) - ord('a')] += 1
            return tuple(arr)

        m = defaultdict(list)
        for s in strs:
            m[key(s)].append(s)
        return list(m.values())


# leetcode submit region end(Prohibit modification and deletion)


# test from here
if __name__ == '__main__':
    print(Solution())
